dace/libraries/blas/nodes/gemv.py

```diff
@@ -83,56 +84,608 @@ def make_sdfg(node, parent_state, parent_sdfg):
-            mul_program, {
-                "__out":
-                dace.Memlet.simple(
-                    mul_out, "__i0", wcr_str="lambda x, y: x + y")
+        state.add_mapped_tasklet("_GEMV_", {
+            "__i%d" % i: "0:%s" % s
+            for i, s in enumerate([N, M])
+        }, {
+            "__A":
+            dace.Memlet(
+                "_A[{}]".format("__i1, __i0" if node.transA else "__i0, __i1")),
+            "__x":
+            dace.Memlet("_x[__i1]")
+        },
+                                 mul_program, {
+                                     "__out":
+                                     dace.Memlet(f"{mul_out}[__i0]",
+                                                 wcr="lambda x, y: x + y")
+                                 },
+                                 external_edges=True,
+                                 output_nodes=output_nodes)
+
+        add_program = "__y_out = ({} * __y_in) + __tmp".format(node.beta)
+
+        memlet_idx = "__i"
+
+        # addition map
+        if node.beta != 0:
+            state.add_mapped_tasklet("_Add_", {"__i": "0:{}".format(N)}, {
+                "__y_in": dace.Memlet(f"_y[{memlet_idx}]"),
+                "__tmp": dace.Memlet(f"{mul_out}[__i]"),
             },
-            external_edges=True,
-            output_nodes=output_nodes)
+                                     add_program,
+                                     {"__y_out": dace.Memlet("_y[__i]")},
+                                     external_edges=True,
+                                     input_nodes={mul_out: access_tmp})
 
-        if node.beta != 0:
-            add_program = "__y_out = ({} * __y_in) + __tmp".format(
-                _cast_to_dtype_str(node.beta, dtype_a))
+        return sdfg
 
-            memlet_idx = "__i"
 
-            # addition map
-            state.add_mapped_tasklet(
-                "_Add_", {"__i": "0:{}".format(N)}, {
-                    "__y_in": dace.Memlet.simple("_y", memlet_idx),
-                    "__tmp": dace.Memlet.simple(mul_out, "__i"),
-                },
-                add_program, {"__y_out": dace.Memlet.simple("_y", "__i")},
-                external_edges=True,
-                input_nodes={mul_out: access_tmp})
+@dace.library.expansion
+class ExpandGemvFpgaAccumulate(ExpandTransformation):
+    """
+    This FPGA-oriented expansion iterates over the input matrix A in simple
+    row-major order, with optional tiling in both dimensions, where the tiles
+    are also traversed in simple row-major order. This means that y is only
+    written once, but x is read for every tile in the y-dimension. The
+    implementation requires accumulation on the output, and does NOT assume
+    native accumulation for the given data type. Instead it uses multiple
+    partial sums to ensure that II=1, and only writes the final accumulated
+    value once it has been combined from the partial sums."""
+    # The above corresponds to gemv_v1 in FBLAS
```

Comment:
Can you say that it also supports A transposed?
Also, I am not sure if we want to indicate how tiles must arrive _when_ this accepts data from a stream (by row if A is non transposed, by column if A is transposed)

Reply:
I've added documentation to both expansions about transposed/non-transposed and vectorization.

Regarding how data must arrive: isn't this covered by "row-major order"? If `A` is not transposed, row-major order means reading sequentially, otherwise it means reading vertically (opposite for the other expansion).